### ISSUE TITLE
Log/StdLog.pm is missing ?

### ISSUE THREAD

**Issue by erdoukki** [robinbowes/net-udap#2]
using github master branch I get error ;

./scripts/udap_shell.pl 
Can't locate Log/StdLog.pm in @INC (you may need to install the Log::StdLog module) (@INC contains:.........) at .../UDAP/net-udap/scripts/../src/Net-UDAP/lib/Net/UDAP/Log.pm line 39.
BEGIN failed--compilation aborted at .../UDAP/net-udap/scripts/../src/Net-UDAP/lib/Net/UDAP/Log.pm line 39.
Compilation failed in require at ./scripts/udap_shell.pl line 31.
BEGIN failed--compilation aborted at ./scripts/udap_shell.pl line 31.

**Comment by hedulduch** [comment]
I have the exact same problem. What to do? Please, help!

**Comment by Demut** [comment]
Has this issue been resolved or am I just as incompetent as I suspect?

**Comment by robinbowes** [comment]
It's been a while since I looked at this code :)

Try installing the following modules before running `./scripts/udap_shell.pl`

`cpan install Log::StdLog Term::Shell Class::Accessor IO::Interface::Simple`

**Comment by Demut** [comment]
Robin,

Thanks I realised after I'd posted that I should have read a little
further, having difficulty installing the modules though, the
IO::Interface::Simple fails with a fatal error on the dll compilation.

I'll persevere, I just want to keep my SBR going a bit longer if I can

Tim

**Comment by Demut** [comment]
Robin,

It seems to be the IO::Interface::Simple that won't install is there an
alternative to cpan to install it?

Tim

**Comment by robinbowes** [comment]
On what platform are you running?

**Comment by Demut** [comment]
Windows 10 64bit - I might give it a go on an old machine I have running XP

**Comment by robinbowes** [comment]
And what perl version/distro? They might have some other way to install modules?

**Comment by Demut** [comment]
ActivePerl and I tried Strawberry Perl, I wonder if it is a path issue, not
finding the right components although I installed complete packages to the
default location from the install packages

**Comment by Demut** [comment]
This is the error I get when trying to install the IO::Interface::Simple module

Created MYMETA.yml and MYMETA.json
Creating new 'Build' script for 'IO-Interface' version '1.09'
Can't find string terminator "'" anywhere before EOF at -e line 1.
  LDS/IO-Interface-1.09.tar.gz
  C:\Perl64\bin\perl.exe Build.PL --installdirs=site -- OK
Running Build for L/LD/LDS/IO-Interface-1.09.tar.gz
Building IO-Interface
C:\Perl64\site\bin\gcc.exe -c -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -DPERL_TEXTMODE_SCRIPTS -DUSE_SITECUSTOMIZE -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2 "-DVERSION=\"1.09\"" "-DXS_VERSION=\"1.09\"" -I"C:\Perl64\lib\CORE" -I"C:\MinGW\x86_64-w64-mingw32\include" -o "lib\IO\Interface.o" "lib\IO\Interface.c"
lib\IO\Interface.xs:16:23: fatal error: sys/ioctl.h: No such file or directory
compilation terminated.
error building dll file from 'lib\IO\Interface.c' at C:/Perl64/lib/ExtUtils/CBuilder/Platform/Windows.pm line 129.
  LDS/IO-Interface-1.09.tar.gz
  C:\Perl64\bin\perl.exe ./Build -- NOT OK

The other modules installed fine

**Comment by ge0thermal** [comment]
Did anyone ever figure out how to fix the issue? I have 0 experience with this and am stuck with the Log/Stdlog.pm error and then followed by the IO/Interface/Simple.pm error

**Comment by Demut** [comment]
To be honest I gave up and built a new player using a RaspberryPi

**Comment by robinbowes** [comment]
I'm grabbing a win10 vagrant box to give this a try...

R.

**Comment by FreeMinded** [comment]
I was also stuck on this error. I have installed `cpan install Log::StdLog Term::Shell Class::Accessor IO::Interface::Simple` successfully on my Ubuntu 16.04 LTS but the error still persisted with the tool from GitHub.
Using the tool from the old repo (https://projects.robinbowes.com/Net-UDAP/trac) it actually works. Meaning I can run it alright (also without installing the modules above). And I was able to make my Receiver work by following the instructions from the Wiki on GitHub (didn't work with the other). Happy Squeezeboxing again! :)

**Comment by robinbowes** [comment]
@FreeMinded It shouldn't be a problem on linux - if you installed the modules, it should work.

**Comment by robinbowes** [comment]
Try the instructions here:

http://forums.slimdevices.com/showthread.php?57861-Net-UDAP-SqueezeBox-Receiver-configuration-tool&p=868057&viewfull=1#post868057

**Comment by FreeMinded** [comment]
@robinbowes Thanks! I just tried again with the GitHub version and the instructions from 
> http://forums.slimdevices.com/showthread.php?57861-Net-UDAP-SqueezeBox-Receiver-configuration-tool&p=868057&viewfull=1#post868057

and indeed it works!
Open Source rules, thanks!

**Comment by hazymat** [comment]
I'm running Win 10 or Server 2016, and I followed the instructions on this forum post:

http://forums.slimdevices.com/showthread.php?57861-Net-UDAP-SqueezeBox-Receiver-configuration-tool&p=868057&viewfull=1#post868057

But still get an error.

When I run udap_shell.pl, I get this:

```
D:\Shared\Installs\net-UDAP\scripts>udap_shell.pl
Can't locate IO/Interface/Simple.pm in @INC (you may need to install the IO::Interface::Simple module) (@INC contains: D:/Shared/Installs/net-UDAP/scripts/../src/Net-UDAP/lib C:/Perl64/site/lib C:/Perl64/lib .) at D:/Shared/Installs/net-UDAP/scripts/../src/Net-UDAP/lib/Net/UDAP.pm line 30.
BEGIN failed--compilation aborted at D:/Shared/Installs/net-UDAP/scripts/../src/Net-UDAP/lib/Net/UDAP.pm line 30.
[snip]
```
So then I run `cpan Log::StdLog Term::Shell Class::Accessor IO::Interface::Simple` but get this:

```
Building IO-Interface
C:\Perl64\site\bin\gcc.exe -c -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -DPERL_TEXTMODE_SCRIPTS -DUSE_SITECUSTOMIZE -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2 "-DVERSION=\"1.09\"" "-DXS_VERSION=\"1.09\"" -I"C:\Perl64\lib\CORE" -I"C:\MinGW\x86_64-w64-mingw32\include" -o "lib\IO\Interface.o" "lib\IO\Interface.c"
lib\\IO\\Interface.xs:16:23: fatal error: sys/ioctl.h: No such file or directory
compilation terminated.
error building dll file from 'lib\IO\Interface.c' at C:/Perl64/lib/ExtUtils/CBuilder/Platform/Windows.pm line 129.
  LDS/IO-Interface-1.09.tar.gz
  C:\Perl64\bin\perl.exe ./Build -- NOT OK
```

I have googled for various parts of this error message, but of course get a load of irrelevant results, mostly pointing me towards documents with troubleshooting that is for Perl on Linux.

Any ideas? Thanks!

**Comment by southseaboy** [comment]
I've been trying to run udap_shell.pl and am encountering the same issues as **@hazymat**.  I've tried a couple of different platforms / varying flavours of perl and in each case I reach a problem with the IO::Interface::Simple module refusing to make.  

For info: I've tried Windows 10 with the latest community ActiveState perl (they won't load the module and their module search shows that the module hasn't successfully compiled for Windows for some time).  I also tried Strawberry perl 5.14.4.1 (the oldest 32 bit version they have available in the recommended downloads section), in this case the module fails to make with various errors in the dependency chain.   I then tried Debian Stretch (in an LXC container on Proxmox) - again make errors, and Debian Wheezy (same). 

I've not got an urgent need to resolve this having found another work-around that meant that I didn't need to change my Duet receivers' config, but would like to feel I had the tool to fall back on whenever I next venture to change network configuration or WiFi passcode. 

Thanks for any suggestions as how to resolve this.

Chris

**Comment by Demut** [comment]
Chris,

I had very much the same problem with Windows 10, both ActiveState perl and
Strawberry perl with my Duet. Ultimately I gave up on the Duet and built a
Rasberry Pi player using piCorePlayer

Tim

**Comment by deezysix** [comment]
I realize this is a very old thread, but in the interest of saving SBRs everywhere...

@hazymat and @southseaboy - Tonight I found myself dealing with your cpan/perl issues, but I was able to use [this page](https://ubuntuforums.org/showthread.php?t=692890) to ultimately get net-udap back up and running (I'm on Ubuntu Server 16.04.4 LTS and it worked fine.)

edit:  Specifically the last (or second to last) comment on the page.  ;)  Cheers.

**Comment by robinbowes** [comment]
It shouldnt be a problem on Ubuntu as long as you have the build tools installed.

That said, im currently investigating a solution using docker. I need to dig out an sbr and check how docker networking functions with the udap broadcast packets and responses.

**Comment by erdoukki** [comment]
new try !
```
gerald@P3530:~/net-udap$ git clone https://github.com/robinbowes/net-udap.git
gerald@P3530:~/net-udap$ cd net-udap/
gerald@P3530:~/net-udap$ sudo cpan Log::StdLog Term::Shell Class::Accessor IO::Interface::Simple
gerald@P3530:~/net-udap$ ./scripts/udap_shell.pl 
UDAP> discover
info: <<< Broadcasting adv_discovery message to MAC address 00:00:00:00:00:00 on 255.255.255.255
info: >>> adv_discovery response received from 00:04:20:17:fc:27
info: <<< Broadcasting get_ip message to MAC address 00:04:20:17:fc:27 on 255.255.255.255
info: >>> get_ip response received from 00:04:20:17:fc:27
info: <<< Broadcasting get_data message to MAC address 00:04:20:17:fc:27 on 255.255.255.255
info: >>> get_data response received from 00:04:20:17:fc:27
UDAP> help
Type 'help command' for more detailed help on a command.
  Commands:
    configure - configure a device
    discover  - Discover UDAP devices and get their current configuration
    exit      - Exit configure mode (if configuring a device), otherwise exit application
    fields    - Display a list of valid device fields
    help      - prints this screen, or help on 'command'
    list      - List discovered devices, or a specific information about a device
    quit      - Exit configure mode (if configuring a device), otherwise exit application
    reset     - Reset a device
    save_data - Save data parameters to device(s)
    save_ip   - Save ip parameters to device(s)
    set       - Set device parameter(s)
UDAP> 
```
THANKS !!

```
gerald@P3530:~$ uname -ar
Linux P3530 5.11.0-35-generic #37-Ubuntu SMP Fri Sep 3 13:59:58 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
gerald@P3530:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 21.04
Release:	21.04
Codename:	hirsute

```

**Comment by joerijuxtra** [comment]
For Windows 10 environment with ActiveState PERL - to avoid the IO error problem:
try downloading the previous version of UDAP v.1.0.x from GitHub and follow the same steps.
This worked for me. The list command for the duet player did not populate all parameters at first, but these can be set/added after which they appear (can be retreived from fields command). It might help to also run the save_ip command beside the save_data command (before reset).
DHCP modus did not want to kick in after reset & power cycling, so I went towards fixed IP config instead (lan_ip_mode=0), which worked. 
Note: the connection via UDAP remained active (inside the Command Promt Window) after I switched from the direct Ethernet cable connection towards Ethernet switch connection for PC & duet device, which helped to further troubleshoot connectivity.
@robinbowes : Thanks for this nice tool to get the control back over the duet player - as my duet controller screen died.
